File: everything_parsing/25-ll1-table/js/ll1table.mjs
// Module 25 — The LL(1) Parsing Table (mirror of java/LL1Table.java).
//
// M[A][t] = the production to apply when expanding A with lookahead t.
// Build: for each A -> alpha, for each t in PREDICT(A -> alpha), set M[A][t].
// A cell with two productions is a CONFLICT <=> grammar is not LL(1).

import { Predict, setStr, END } from "./predict.mjs";

export class LL1Table {
  constructor(g) {
    this.g = g;
    this.pr = new Predict(g);

    const cols = new Set(g.terminals());
    cols.add(END);
    this.columns = [...cols];

    this.M = new Map();
    for (const nt of g.nonterminals) this.M.set(nt, new Map());
    for (const p of g.productions)
      for (const t of this.pr.predict(p)) {
        const row = this.M.get(p.lhs);
        if (!row.has(t)) row.set(t, []);
        row.get(t).push(p);
      }
  }

  cell(nt, t) {
    const row = this.M.get(nt);
    return row && row.has(t) ? row.get(t) : [];
  }

  conflicts() {
    const out = [];
    for (const nt of this.g.nonterminals)
      for (const [tok, ps] of this.M.get(nt))
        if (ps.length > 1) out.push({ nt, token: tok, ps });
    return out;
  }

  isLL1() { return this.conflicts().length === 0; }

  render() {
    let w = 4;
    for (const c of this.columns) w = Math.max(w, c.length + 1);

    let sb = pad("", 6);
    for (const c of this.columns) sb += pad(c, w);
    sb += "\n";
    for (const nt of this.g.nonterminals) {
      sb += pad(nt, 6);
      for (const c of this.columns) {
        const ps = this.cell(nt, c);
        const txt = ps.length === 0 ? "." : ps.length === 1 ? String(ps[0].index) : "!";
        sb += pad(txt, w);
      }
      sb += "\n";
    }
    return sb;
  }
}

export function compact(p) {
  return "(" + p.index + ") " + p.lhs + " -> " + (p.rhs.length === 0 ? "epsilon" : p.rhs.join(" "));
}

function pad(s, w) {
  while (s.length < w) s += " ";
  return s;
}

File: everything_parsing/28-grammar-tool-project2/java/LL1Table.java
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Module 25 — The LL(1) Parsing Table.
 *
 * M[A][t] = the one production to apply when the parser is expanding
 * nonterminal A and the lookahead token is t.
 *
 * Build rule: for each production p = A -> alpha, for each terminal t in
 * PREDICT(p), set M[A][t] = p. If M[A][t] is already set, that cell is a
 * CONFLICT and the grammar is not LL(1) — the two productions competing for the
 * cell are a concrete witness.
 *
 * PREDICT / FIRST / FOLLOW come from Module 20's Predict.
 */
public final class LL1Table {

    public final Grammar g;
    public final Predict pr;
    public final List<String> columns;                        // terminals + $
    public final Map<String, Map<String, List<Grammar.Production>>> M = new LinkedHashMap<>();

    public LL1Table(Grammar g) {
        this.g = g;
        this.pr = new Predict(g);

        Set<String> cols = new LinkedHashSet<>(g.terminals());
        cols.add(Predict.END);
        this.columns = new ArrayList<>(cols);

        for (String nt : g.nonterminals) M.put(nt, new LinkedHashMap<>());
        for (Grammar.Production p : g.productions)
            for (String t : pr.predict(p))
                M.get(p.lhs()).computeIfAbsent(t, k -> new ArrayList<>()).add(p);
    }

    public List<Grammar.Production> cell(String nt, String t) {
        return M.getOrDefault(nt, Map.of()).getOrDefault(t, List.of());
    }

    public record Conflict(String nt, String token, List<Grammar.Production> ps) {}

    public List<Conflict> conflicts() {
        List<Conflict> out = new ArrayList<>();
        for (String nt : g.nonterminals)
            for (var e : M.get(nt).entrySet())
                if (e.getValue().size() > 1) out.add(new Conflict(nt, e.getKey(), e.getValue()));
        return out;
    }

    public boolean isLL1() { return conflicts().isEmpty(); }

    /**
     * Grid render: one row per nonterminal, one column per terminal (+ $).
     * Cells hold the production INDEX (see the numbered legend printed above);
     * "." = blank (a syntax error at parse time), "!" = conflict.
     */
    public String render() {
        int w = 4;
        for (String c : columns) w = Math.max(w, c.length() + 1);

        StringBuilder sb = new StringBuilder();
        sb.append(pad("", 6));
        for (String c : columns) sb.append(pad(c, w));
        sb.append('\n');
        for (String nt : g.nonterminals) {
            sb.append(pad(nt, 6));
            for (String c : columns) {
                List<Grammar.Production> ps = cell(nt, c);
                String txt = ps.isEmpty() ? "." :
                    ps.size() == 1 ? Integer.toString(ps.get(0).index()) : "!";
                sb.append(pad(txt, w));
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    static String compact(Grammar.Production p) {
        return "(" + p.index() + ") " + p.lhs() + " -> "
             + (p.rhs().isEmpty() ? "epsilon" : String.join(" ", p.rhs()));
    }
    private static String pad(String s, int w) {
        StringBuilder b = new StringBuilder(s);
        while (b.length() < w) b.append(' ');
        return b.toString();
    }
}
